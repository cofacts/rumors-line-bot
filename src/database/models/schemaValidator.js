import Ajv from 'ajv';

const ajv = new Ajv();

const SCHEMAS = {
  userArticleLink: require('./userArticleLink.json'),
  userSettings: require('./userSettings.json'),
  appVariable: require('./appVariable.json'),
};

const ClassTable = {
  Date: Date,
};

ajv.addKeyword('instanceOf', {
  compile: function (className) {
    var Class = ClassTable[className];
    return function (data) {
      return data instanceof Class;
    };
  },
});

export function compile(schemaName) {
  return (data) => {
    const validate = ajv.compile(SCHEMAS[schemaName]);
    const valid = validate(data);
    return { valid, errors: validate.errors };
  };
}

export const validators = {
  userArticleLink: compile('userArticleLink'),
  userSettings: compile('userSettings'),
  appVariable: compile('appVariable'),
};
