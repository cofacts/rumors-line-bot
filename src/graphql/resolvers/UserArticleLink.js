export default {
  /**
   * @param {UserArticleLink} source
   */
  article(source) {
    // Returns stub
    return { id: source.articleId };
  },
};
