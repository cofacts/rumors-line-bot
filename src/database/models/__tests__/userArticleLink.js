import Client from '../../mongoClient';
import MockDate from 'mockdate';
import { validators } from '../schemaValidator';
import UserArticleLink from '../userArticleLink';

const userArticleLinkValidator = validators.userArticleLink;

const FIXED_DATE = 612921600000;

describe('userArticleLink', () => {
  beforeAll(async () => {
    MockDate.set(FIXED_DATE);
  });

  beforeEach(async () => {
    if (await UserArticleLink.collectionExists()) {
      await (await UserArticleLink.client).drop();
    }
  });

  afterAll(async () => {
    await (await Client.getInstance()).close();
    MockDate.reset();
  });

  it('[schema] should pass', async () => {
    const data = {
      userId: 'this_is_user_id',
      articleId: 'this_is_article_id',
      createdAt: new Date(),
    };
    const result = userArticleLinkValidator(data);
    expect(result).toMatchSnapshot();
  });

  it('[schema] should fail with wrong type of createdAt', async () => {
    const data = {
      userId: 'this_is_user_id',
      articleId: 'this_is_article_id',
      createdAt: '2020-01-01T20:10:18.314Z',
    };
    const result = userArticleLinkValidator(data);
    expect(result).toMatchSnapshot();
  });

  it('[model] should create a document', async () => {
    const result = await UserArticleLink.create({
      userId: 'this_is_user_id',
      articleId: 'this_is_article_id',
    });
    delete result._id;
    expect(result).toMatchSnapshot();
  });

  it('[model] should fail to create a document', async () => {
    expect(
      UserArticleLink.create({
        articleId: 'this_is_article_id',
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('[model] findOrInsertByUserIdAndArticleId()', async () => {
    const userId = 'userId-0';
    const articleId = 'articleId-0';

    await UserArticleLink.create({ userId, articleId });

    const result = await UserArticleLink.findOrInsertByUserIdAndArticleId(
      userId,
      articleId
    );
    delete result._id;

    expect(result).toMatchSnapshot();
  });

  it('[model] findOrInsertByUserIdAndArticleId() (upsert)', async () => {
    const userId = 'userId-1';
    const articleId = 'articleId-1';

    const result = await UserArticleLink.findOrInsertByUserIdAndArticleId(
      userId,
      articleId
    );
    delete result._id;

    expect(result).toMatchSnapshot();
  });

  it('[model] updateTimestamps()', async () => {
    const userId = 'userId-2';
    const articleId = 'articleId-2';

    await UserArticleLink.create({ userId, articleId });

    const updatedDate = await UserArticleLink.updateTimestamps(
      userId,
      articleId,
      {
        lastViewedAt: new Date(FIXED_DATE + 60 * 1000),
      }
    );

    delete updatedDate._id;
    expect(updatedDate).toMatchSnapshot();
  });

  it('[model] updateTimestamps() (upsert)', async () => {
    const userId = 'userId-3';
    const articleId = 'articleId-3';

    const updatedData = await UserArticleLink.updateTimestamps(
      userId,
      articleId,
      {
        lastViewedAt: new Date(FIXED_DATE + 60 * 1000),
      }
    );

    delete updatedData._id;
    expect(updatedData).toMatchSnapshot();
  });
});
