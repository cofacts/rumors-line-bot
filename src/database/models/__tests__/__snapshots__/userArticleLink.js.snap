// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userArticleLink [model] findByUserId() should return data 1`] = `
Array [
  Object {
    "_id": "_id",
    "articleId": "articleId-1",
    "createdAt": 1989-06-04T00:00:00.000Z,
    "userId": "userId-4",
  },
  Object {
    "_id": "_id",
    "articleId": "articleId-2",
    "createdAt": 1989-06-04T00:00:00.000Z,
    "userId": "userId-4",
  },
]
`;

exports[`userArticleLink [model] findOrInsertByUserIdAndArticleId() (upsert) 1`] = `
Object {
  "articleId": "articleId-1",
  "createdAt": 1989-06-04T00:00:00.000Z,
  "userId": "userId-1",
}
`;

exports[`userArticleLink [model] findOrInsertByUserIdAndArticleId() 1`] = `
Object {
  "articleId": "articleId-0",
  "createdAt": 1989-06-04T00:00:00.000Z,
  "userId": "userId-0",
}
`;

exports[`userArticleLink [model] should create a document 1`] = `
Object {
  "articleId": "this_is_article_id",
  "createdAt": 1989-06-04T00:00:00.000Z,
  "userId": "this_is_user_id",
}
`;

exports[`userArticleLink [model] should fail to create a document 1`] = `
"[
  {
    \\"keyword\\": \\"required\\",
    \\"dataPath\\": \\"\\",
    \\"schemaPath\\": \\"#/required\\",
    \\"params\\": {
      \\"missingProperty\\": \\"userId\\"
    },
    \\"message\\": \\"should have required property 'userId'\\"
  }
]"
`;

exports[`userArticleLink [model] updateTimestamps() (upsert) 1`] = `
Object {
  "articleId": "articleId-3",
  "createdAt": 1989-06-04T00:00:00.000Z,
  "lastViewedAt": 1989-06-04T00:01:00.000Z,
  "userId": "userId-3",
}
`;

exports[`userArticleLink [model] updateTimestamps() 1`] = `
Object {
  "articleId": "articleId-2",
  "createdAt": 1989-06-04T00:00:00.000Z,
  "lastViewedAt": 1989-06-04T00:01:00.000Z,
  "userId": "userId-2",
}
`;

exports[`userArticleLink [schema] should fail with wrong type of createdAt 1`] = `
Object {
  "errors": Array [
    Object {
      "dataPath": ".createdAt",
      "keyword": "instanceOf",
      "message": "should pass \\"instanceOf\\" keyword validation",
      "params": Object {
        "keyword": "instanceOf",
      },
      "schemaPath": "#/properties/createdAt/instanceOf",
    },
  ],
  "valid": false,
}
`;

exports[`userArticleLink [schema] should pass 1`] = `
Object {
  "errors": null,
  "valid": true,
}
`;
